examples/nosecone-nextjs-router/middleware.ts

```diff
@@ -0,0 +1,75 @@
+import { type NoseconeOptions, defaults, createMiddleware as noseconeMiddleware } from "@nosecone/next";
+import {
+    type NextFetchEvent,
+    type NextMiddleware,
+    type NextRequest,
+    NextResponse,
+} from "next/server";
+import { match } from "path-to-regexp";
+
+// Next.js middleware config
+export const config = {
+    matcher: ['/((?!_next/|_static|_vercel|[\\w-]+\\.\\w+).*)'],
+};
+
+// Nosecone security headers configuration
+// https://docs.arcjet.com/nosecone/quick-start
+const noseconeOptions: NoseconeOptions = {
+    ...defaults,
+    // Customize security headers here
+    // See https://docs.arcjet.com/nosecone/reference
+};
+const securityHeaders = noseconeMiddleware(noseconeOptions);
+
+// Add any paths you want to run different middleware for. They use
+// path-to-regexp which is the same as the Next.js config. You can provide a
+// single middleware or an array of middlewares.
+export default router({
+    // Run nosecone middleware on any path
+    "/{*path}": [securityHeaders]
+});
+
+// A simple middleware router that allows you to run different middleware based
+// on the path of the request.
+function router(
+    pathMiddlewareMap: Record<string, NextMiddleware | NextMiddleware[]>,
+): NextMiddleware {
+    const middleware = Object.entries(pathMiddlewareMap).map(
+        ([path, middleware]) => {
+            if (Array.isArray(middleware)) {
+                return [match(path), middleware] as const;
+            } else {
+                return [match(path), [middleware]] as const;
+            }
+        },
+    );
+
+    return async (
+        request: NextRequest,
+        event: NextFetchEvent,
+    ): Promise<NextResponse | Response> => {
+        const path = request.nextUrl.pathname || "/";
+        const addedHeaders = new Headers();
+
+        for (const [matchFunc, middlewareFuncs] of middleware) {
+            const m = matchFunc(path);
+            if (m) {
+                for (const fn of middlewareFuncs) {
+                    const resp = await fn(request, event);
+                    // TODO: better response guards
```

Comment:
@blaine-arcjet could you review my fix in https://github.com/arcjet/arcjet-js/issues/2639#issuecomment-2558514321 please? Seems like we should handle redirects and auth errors by just returning the response, but that might leave the security headers unset.

Reply:
Is there a reason we can't add the headers to returned response before returning it from the middleware?

It looks like the implementation for `NextResponse.redirect()` isn't complex:
```js
static redirect(url, init) {
        const status = typeof init === "number" ? init : (init == null ? void 0 : init.status) ?? 307;
        if (!REDIRECTS.has(status)) {
            throw new RangeError('Failed to execute "redirect" on "response": Invalid status code');
        }
        const initObj = typeof init === "object" ? init : {};
        const headers = new Headers(initObj == null ? void 0 : initObj.headers);
        headers.set("Location", (0, _utils.validateURL)(url));
        return new NextResponse(null, {
            ...initObj,
            headers,
            status
        });
}
```

So we could return a `Response` instantly that sets the status code & Location header.

Also, to make this resilient, we probably want to handle `NextResponse.rewrite()`, which looks like it sets a different header than `NextResponse.next()`:

```js
static rewrite(destination, init) {
        const headers = new Headers(init == null ? void 0 : init.headers);
        headers.set("x-middleware-rewrite", (0, _utils.validateURL)(destination));
        handleMiddlewareField(init, headers);
        return new NextResponse(null, {
            ...init,
            headers
        });
}
```

Reply:
Ideally we'd add the headers to all responses. So this is really about extending how we handle the redirect and rewrites to make sure they're captured correctly.